### PR TITLE
HTTP logging middleware in .NET 8

### LINKED ISSUE #30642: HTTP logging middleware has several new capabilities:
Also add to WN doc.

### Description

See [HTTP logging](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-rc-2/#http-logging-extensibility)

### Page URL

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-logging/?view=aspnetcore-7.0

### Content source URL

https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/fundamentals/http-logging/index.md

### Version Independent Document ID

b39594cc-9d1c-611c-e3e7-8755d74253a6

### Article author

Rick-Anderson

---
[Associated WorkItem - 169119](https://dev.azure.com/msft-skilling/Content/_workitems/edit/169119)

### PR DESCRIPTION
**Pull request by tdykstra**
Fixes #30642 

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/fundamentals/http-logging/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/c903c3ff93770294120e50b0fd18288e4a1e187d/aspnetcore/fundamentals/http-logging/index.md) | [HTTP Logging in ASP.NET Core](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/http-logging/index?branch=pr-en-us-30814) |

### CODE REVIEW

**Review by Rick-Anderson — APPROVED** [review]

File: aspnetcore/fundamentals/http-logging/index.md [hunk 1]
```diff
@@ -1,71 +1,94 @@
...
+<xref:Microsoft.AspNetCore.HttpLogging.HttpLoggingOptions.RequestHeaders> and <xref:Microsoft.AspNetCore.HttpLogging.HttpLoggingOptions.ResponseHeaders> are sets of HTTP headers that are logged. Header values are only logged for header names that are in these collections. The following code adds `sec-ch-ua` to the <xref:Microsoft.AspNetCore.HttpLogging.HttpLoggingOptions.RequestHeaders>, so the value of the `sec-ch-ua` header is logged. And it adds `MyResponseHeader` to the <xref:Microsoft.AspNetCore.HttpLogging.HttpLoggingOptions.ResponseHeaders>, so the value of the `MyResponseHeader` header is logged. If these lines are removed, the values of these headers are redacted.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/http-logging/index.md:95]:
```suggestion
<xref:Microsoft.AspNetCore.HttpLogging.HttpLoggingOptions.RequestHeaders> and <xref:Microsoft.AspNetCore.HttpLogging.HttpLoggingOptions.ResponseHeaders> are sets of HTTP headers that are logged. Header values are only logged for header names that are in these collections. The following code adds `sec-ch-ua` to the <xref:Microsoft.AspNetCore.HttpLogging.HttpLoggingOptions.RequestHeaders>, so the value of the `sec-ch-ua` header is logged. And it adds `MyResponseHeader` to the <xref:Microsoft.AspNetCore.HttpLogging.HttpLoggingOptions.ResponseHeaders>, so the value of the `MyResponseHeader` header is logged. If these lines are removed, the values of these headers are `[Redacted]`.
```

File: aspnetcore/fundamentals/http-logging/index.md [hunk 2]
```diff
@@ -90,6 +113,71 @@ This approach can also be used to enable logging for data that is not logged by
...
+## `IHttpLoggingInterceptor`
+
+<xref:Microsoft.AspNetCore.HttpLogging.IHttpLoggingInterceptor> is the interface for a service that can be implemented to handle per-request and per-response callbacks for customizing what details get logged. Any endpoint-specific log settings are applied first and can then be overridden in these callbacks. An implementation can:
+
+* Inspect a request or response.
+* Enable or disable any <xref:Microsoft.AspNetCore.HttpLogging.HttpLoggingFields>.
+* Adjust how much of the request or response body is logged.
+* Add custom fields to the logs.
+
+Register an `IHttpLoggingInterceptor` implementation by calling <xref:Microsoft.Extensions.DependencyInjection.HttpLoggingServicesExtensions.AddHttpLogging%2A> in `Program.cs`. If multiple `IHttpLoggingInterceptor` instances are registered, they're run in the order registered.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/http-logging/index.md:135]: `AddHttpLoggingInterceptor<T>`

File: aspnetcore/fundamentals/http-logging/samples/8.x/SampleHttpLoggingInterceptor.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
+using Microsoft.AspNetCore.HttpLogging;
+
+namespace HttpLoggingSample;
+
+internal sealed class SampleHttpLoggingInterceptor : IHttpLoggingInterceptor
+{
+    public ValueTask OnRequestAsync(HttpLoggingInterceptorContext logContext)
+    {
+        if (!logContext.HttpContext.Request.Path.StartsWithSegments("/api"))
```
Comment (Tratcher) [inline aspnetcore/fundamentals/http-logging/samples/8.x/SampleHttpLoggingInterceptor.cs:9]: Sorry, I know this was copied from my own sample, but directly matching the path like this is a bad habit we don't want to encourage. Can you change this to exclude logs for POST requests instead?

**Review by Tratcher — COMMENTED** [review]
We also need coverage for HttpLoggingAttribute and the relative priority of options. HttpLoggingOptions apply first, can be overriden by HttpLoggingAttribute per request/endpoint, and then IHttpLoggingInterceptor is called with the results and can further modify them per request.

File: aspnetcore/fundamentals/http-logging/index.md [hunk 4]
```diff
@@ -90,6 +119,85 @@ This approach can also be used to enable logging for data that is not logged by
...
+
+Register an `IHttpLoggingInterceptor` implementation by calling [`AddHttpLoggingInterceptor<T>`](xref:Microsoft.Extensions.DependencyInjection.HttpLoggingServicesExtensions.AddHttpLoggingInterceptor%60%601(Microsoft.Extensions.DependencyInjection.IServiceCollection)) in `Program.cs`. If multiple `IHttpLoggingInterceptor` instances are registered, they're run in the order registered.
+
+The following example shows how to register an `IHttpLoggingInterceptor` implementation:
+
+[!code-csharp[](~/fundamentals/http-logging/samples/8.x/Program.cs?name=snippet4&highlight=7)]
+
+The following example is an `IHttpLoggingInterceptor` implementation that:
+
+* Inspects the request path and disables logging for POST requests.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/http-logging/index.md:151]:
```suggestion
* Inspects the request method and disables logging for POST requests.
```

File: aspnetcore/fundamentals/http-logging/samples/8.x/Program.cs [hunk 5]
```diff
@@ -59,5 +59,50 @@
...
+    await next();
+});
+
+app.MapGet("/", () => "Hello World!");
+
+// <snippet5>
+app.MapGet("/duration", [HttpLogging(loggingFields: HttpLoggingFields.Duration)]
+    () => "Hello World! (logging duration)");
+
+app.MapGet("/request", [HttpLogging(loggingFields: HttpLoggingFields.RequestPropertiesAndHeaders)]
```
Comment (Tratcher) [inline aspnetcore/fundamentals/http-logging/samples/8.x/Program.cs:102]: Oh yeah, we also have `.WithHttpLogging(HttpLoggingFields.RequestPropertiesAndHeaders);` for this minimal scenario. The attribute is more for controllers, but it works here too.
https://github.com/dotnet/aspnetcore/blob/82ff9246592477abf4ff57ac905a368cbd99ff20/src/Middleware/HttpLogging/src/HttpLoggingEndpointConventionBuilderExtensions.cs#L23

**Review by Tratcher — APPROVED** [review]

**Review by iamcymentho — APPROVED** [review]

**Pull request merged**